include/posix/sys/types.h

```diff
@@ -35,10 +35,15 @@ typedef struct k_sem sem_t;
 
 /* Mutex */
 typedef struct pthread_mutex {
+	struct k_mem_block block;
 	struct k_sem *sem;
+	pthread_t owner;
+	u32_t lock_count;
+	int type;
```

Comment:
This changes makes struct pthread_mutex **six times larger**.  Is that really worth it for the features being added?  Is there no way to configure out the stuff you don't want (e.g. a kconfig that allows you to choose whether or not to support recursive locks, which would eliminate owner and count...)

Reply:
We can remove k_block and replace sem pointer with sem object i.e. struct k_sem sem; Also u32_t lock_count to u16_t lock_count. Would that be fine ? @andyross  I will fix other comments as well.

Reply:
You could pack type into the same word as lock_count too.  I'd be... OK with that I guess.  But it still has the effect of making a pthread mutex a **much** heavier abstraction than the underlying semaphore, for IMHO not much benefit in typical code.  (Let's be honest: "no one" actually uses recursive mutexes in real world code, they're an obscure feature specified to meet someone's checklist, which makes them a poor fit for Zephyr).   Ideally I'd like the overhead to be eliminatable via kconfig and off by default.